File: experiment/static/graph.js
/* A stable, keyboard-accessible neighborhood, with dependency arrows behind it. */
(() => {
  const states = {
    building: "Building",
    excluded: "Excluded",
    available: "Available",
    failed: "Failed",
    blocked: "Blocked",
    waiting: "Waiting / not started",
    settling: "Awaiting result",
    unknown: "Not observed",
  };
  const phases = {
    configurePhase: "Configuring",
    buildPhase: "Building",
    checkPhase: "Running checks",
    installCheckPhase: "Running install checks",
    installPhase: "Installing",
    fixupPhase: "Finalizing",
    unpackPhase: "Unpacking",
    patchPhase: "Patching",
  };
  let campaign = null,
    focus = null,
    page = 0,
    generation = 0,
    current = null,
    signature = null;
  const name = (n) =>
    n.labels[0]?.label ||
    n.name.replaceAll("-x86_64-unknown-linux-gnufilc0", "");
  const phase = (n) =>
    n.state === "building"
      ? phases[n.phase] || n.phase || states[n.state]
      : states[n.state];
  function choose(drv, fromRoute = false) {
    if (window.dashboardNavigation && !fromRoute)
      return window.dashboardNavigation.focus(drv);
    focus = drv;
    page = 0;
    signature = null;
    refreshGraph();
  }
  function node(n, center = false) {
    const b = el(
      "button",
      null,
      "graph-node " + n.state + (center ? " focal" : ""),
    );
    b.dataset.drv = n.drv;
    b.title = n.name + "\n" + n.drv;
    if (n.availability_evidence === "consumer-phase")
      b.title +=
        "\nAvailability inferred from a consumer reaching a build phase. Build/test origin remains unknown.";
    b.append(
      el("span", name(n), "graph-node-name"),
      el("span", phase(n), "graph-node-state"),
    );
    if (n.roles?.length)
      b.append(el("small", n.roles.join(" · "), "graph-role"));
    if (n.failure) b.append(el("small", n.failure, "graph-role"));
    b.setAttribute(
      "aria-label",
      `${name(n)}: ${phase(n)}. ${center ? "Open details" : "Explore dependencies"}`,
    );
    b.onclick = () => (center ? showDerivation(n.drv) : choose(n.drv));
    return b;
  }
  function chip(n) {
    const b = el(
      "button",
      name(n),
      "graph-chip " +
        n.state +
        (current?.focus?.drv === n.drv ? " selected" : ""),
    );
    b.title = n.name + " · " + states[n.state];
    b.onclick = () => choose(n.drv);
    return b;
  }
  function drawEdges() {
    const svg = $("graph-edges"),
      stage = $("graph-stage"),
      bounds = stage.getBoundingClientRect();
    svg.replaceChildren();
    svg.setAttribute("viewBox", `0 0 ${bounds.width} ${bounds.height}`);
    const ns = "http://www.w3.org/2000/svg";
    const defs = document.createElementNS(ns, "defs"),
      marker = document.createElementNS(ns, "marker");
    for (const [k, v] of Object.entries({
      id: "graph-arrow",
      viewBox: "0 0 10 10",
      refX: "9",
      refY: "5",
      markerWidth: "5",
      markerHeight: "5",
      orient: "auto-start-reverse",
    }))
      marker.setAttribute(k, v);
    const tip = document.createElementNS(ns, "path");
    tip.setAttribute("d", "M 0 0 L 10 5 L 0 10 z");
    tip.setAttribute("fill", "#93a38d");
    marker.append(tip);
    defs.append(marker);
    svg.append(defs);
    const nodes = [...stage.querySelectorAll("[data-drv]")];
    const edges = [...(current?.edges || [])];
    if (nodes.some((n) => n.dataset.drv === "available-inputs"))
      edges.push({ source: "available-inputs", target: current.focus.drv });
    for (const e of edges) {
      const a = nodes.find((n) => n.dataset.drv === e.source),
        b = nodes.find((n) => n.dataset.drv === e.target);
      if (!a || !b) continue;
      const ra = a.getBoundingClientRect(),
        rb = b.getBoundingClientRect(),
        mobile = innerWidth <= 650;
      const x1 = (mobile ? ra.left + ra.width / 2 : ra.right) - bounds.left,
        y1 = (mobile ? ra.bottom : ra.top + ra.height / 2) - bounds.top;
      const x2 = (mobile ? rb.left + rb.width / 2 : rb.left) - bounds.left,
        y2 = (mobile ? rb.top : rb.top + rb.height / 2) - bounds.top;
      const left1 = ra.left - bounds.left,
        left2 = rb.left - bounds.left;
      const middle1 = ra.top + ra.height / 2 - bounds.top,
        middle2 = rb.top + rb.height / 2 - bounds.top;
      const bus = Math.max(2, Math.min(left1, left2) - 14);
      const d = mobile
        ? `M${left1},${middle1} H${bus} V${middle2} H${left2}`
        : `M${x1},${y1} C${(x1 + x2) / 2},${y1} ${(x1 + x2) / 2},${y2} ${x2},${y2}`;
      const path = document.createElementNS(ns, "path");
      path.setAttribute("d", d);
      path.setAttribute("marker-end", "url(#graph-arrow)");
      svg.append(path);
    }
  }
  function renderGraph(g) {
    current = g;
    const fresh = g.heartbeat && Date.now() / 1000 - g.heartbeat < 30;
    $("graph-freshness").textContent = fresh
      ? "Updated just now"
      : "Controller stale · last known state";
    const w = g.work;
    $("graph-work-status").textContent =
      w.kind === "plan"
        ? `Evaluating recipes · ${w.completed} of ${w.total} complete`
        : w.kind === "build"
          ? w.stage === "preflight"
            ? "Checking which outputs are already in the store…"
            : g.building.length
              ? `${g.building.length} build${g.building.length === 1 ? "" : "s"} running in Nix`
              : "Nix is resolving dependencies or fetching outputs"
          : g.mode === "paused"
            ? "Campaign paused"
            : "Between batches";
    if (g.planning && w.kind === "build")
      $("graph-work-status").textContent +=
        ` · planning ${g.planning.completed}/${g.planning.total}`;
    $("graph-batch-label").textContent =
      g.batches?.length > 1
        ? `${g.batches.length} active batches`
        : g.batch?.state === "finished"
          ? "Last build batch"
          : "Batch";
    $("graph-follow").textContent = focus ? "Follow live" : "● Live";
    $("graph-follow").setAttribute("aria-pressed", String(!focus));
    $("graph-back").disabled = !focus;
    const nextSignature = JSON.stringify([
      g.focus,
      g.inputs,
      g.consumers,
      g.roots,
      g.building,
      g.totals,
      g.selected_dependents,
      g.affected_roots,
      g.page,
      focus,
    ]);
    if (nextSignature === signature) return;
    signature = nextSignature;
    const focused = document.activeElement?.dataset?.drv;
    replace(
      "graph-running",
      g.building.map((n) => {
        const b = el("button", name(n) + " · " + phase(n), "graph-live-build");
        b.onclick = () => choose(n.drv);
        return b;
      }),
    );
    replace("graph-roots", g.roots.map(chip));
    $("graph-input-label").textContent = `DEPENDS ON · ${g.totals.inputs || 0}`;
    $("graph-consumer-label").textContent =
      `NEEDED BY · ${g.totals.consumers || 0}`;
    const emptyInputs =
      g.totals.hidden_available === g.totals.inputs && g.totals.inputs
        ? `${g.totals.inputs} available inputs hidden`
        : "No inputs on this page";
    replace(
      "graph-inputs",
      g.inputs.length
        ? g.inputs.map((n) => node(n))
        : [el("p", emptyInputs, "graph-empty")],
    );
    if (g.totals.hidden_available) {
      const ready = el("button", null, "graph-node available graph-collapsed");
      ready.dataset.drv = "available-inputs";
      ready.append(
        el(
          "span",
          `${g.totals.hidden_available} available inputs`,
          "graph-node-name",
        ),
        el("span", "Expand dependencies ↗", "graph-node-state"),
      );
      ready.onclick = () => {
        $("graph-available").checked = true;
        page = 0;
        signature = null;
        refreshGraph();
      };
      if (!g.inputs.length) $("graph-inputs").replaceChildren(ready);
      else $("graph-inputs").append(ready);
    }
    replace(
      "graph-consumers",
      g.consumers.length
        ? g.consumers.map((n) => node(n))
        : [el("p", "No recorded consumers here.", "graph-empty")],
    );
    const center = g.focus
      ? [node(g.focus, true)]
      : [
          el(
            "p",
            "The map will appear as recipes are evaluated.",
            "graph-empty",
          ),
        ];
    if (g.focus) {
      center.push(
        el(
          "p",
          `${g.selected_dependents} selected package${g.selected_dependents === 1 ? " depends" : "s depend"} on this`,
          "graph-impact",
        ),
      );
      if (g.affected_roots.length) {
        center.push(
          el("span", "Leads to these batch targets", "graph-target-label"),
        );
        const list = el("div", null, "graph-roots graph-affected");
        list.append(...g.affected_roots.map(chip));
        center.push(list);
      }
      if (g.focus.attempt) {
        const log = el("button", "Open build log ↗", "graph-log");
        log.onclick = () => showLog(g.focus.attempt, g.focus.drv);
        center.push(log);
      }
    }
    replace("graph-focus", center);
    $("graph-previous").hidden = g.page === 0;
    $("graph-next").hidden = !g.totals.more;
    $("graph-explanation").textContent = g.totals.hidden_available
      ? `${g.totals.hidden_available} ready inputs hidden`
      : "";
    if (focused)
      [...$("graph-stage").querySelectorAll("[data-drv]")]
        .find((n) => n.dataset.drv === focused)
        ?.focus({ preventScroll: true });
    requestAnimationFrame(drawEdges);
  }
  async function refreshGraph() {
    if (!campaign) return;
    const g = ++generation;
    try {
      const q = new URLSearchParams({
        campaign,
        focus: focus || "",
        available: $("graph-available").checked ? "1" : "0",
        page,
      });
      const response = await fetch("/api/graph?" + q);
      if (!response.ok) throw new Error("Graph unavailable");
      const value = await response.json();
      if (g !== generation) return;
      page = value.page;
      renderGraph(value);
    } catch (e) {
      if (g === generation)
        $("graph-freshness").textContent =
          "Connection lost · showing last observation";
    }
  }
  window.dependencyMap = {
    updateCampaign(id) {
      if (id === campaign) return;
      campaign = id;
      focus = null;
      page = 0;
      signature = null;
      refreshGraph();
    },
    focus: (drv) => {
      if (focus !== drv) choose(drv, true);
    },
    locate(drv) {
      window.dashboardNavigation.focus(drv);
    },
  };
  $("graph-follow").onclick = () => window.dashboardNavigation.focus(null);
  $("graph-back").onclick = () => window.dashboardNavigation.backGraph();
  $("graph-available").onchange = () => {
    page = 0;
    signature = null;
    refreshGraph();
  };
  $("graph-previous").onclick = () => {
    page = Math.max(0, page - 1);
    refreshGraph();
  };
  $("graph-next").onclick = () => {
    page++;
    refreshGraph();
  };
  new ResizeObserver(drawEdges).observe($("graph-stage"));
  window.addEventListener("resize", drawEdges);
  window.dependencyMap.updateCampaign(data.campaign?.id);
  setInterval(refreshGraph, 5000);
})();

File: experiment/graph.py
"""A bounded, live neighborhood of the recorded build graph.

This is an observation layer, not another scheduler. A stopped Nix activity does
not imply success, and edges describe required outputs rather than every output
the dependency can produce. No Nix subprocesses or database writes happen here.
"""

from collections import Counter
import json

from .attempt import directory
from .model import stamp


SCOPE = """WITH RECURSIVE scope(drv) AS (
  SELECT drv FROM candidates WHERE campaign=? AND drv IS NOT NULL
  UNION SELECT child FROM edges JOIN scope ON parent=scope.drv)
"""


def attempt_file(state, aid, name):
    path = directory(state, aid) / name
    try:
        with path.open("rb") as f:
            data = f.read(8 * 1024**2 + 1)
        if len(data) > 8 * 1024**2:
            return None
        return json.loads(data)
    except (OSError, ValueError):
        return None


def live_graph(db, state, campaign, focus=None, show_available=False, page=0):
    c = db.execute(
        "SELECT mode,heartbeat FROM campaigns WHERE id=?", (campaign,)
    ).fetchone()
    if not c:
        raise ValueError("unknown campaign")
    active_rows = db.execute(
        "SELECT * FROM attempts WHERE campaign=? AND state IN ('intended','running') ORDER BY kind='build' DESC,created DESC",
        (campaign,),
    ).fetchall()
    active = active_rows[0] if active_rows else None
    planner = next((r for r in active_rows if r["kind"] == "plan"), None)
    batch = db.execute(
        "SELECT * FROM attempts WHERE campaign=? AND kind='build' ORDER BY created DESC LIMIT 1",
        (campaign,),
    ).fetchone()
    builds = [r for r in active_rows if r["kind"] == "build"]
    batches = builds or ([batch] if batch else [])
    if builds:
        batch = builds[0]
    targets = list(dict.fromkeys(d for r in batches for d in json.loads(r["targets"])))
    in_flight = bool(builds)
    observations = [attempt_file(state, r["id"], "before.json") for r in builds]
    before = next((v for v in observations if v is not None), None)
    existing = set(p for v in observations for p in (v or []))
    activities = {}
    provided, requested = set(), set()
    for current in builds:
        activities.update(
            {
                r["drv"]: dict(r)
                for r in db.execute(
                    "SELECT * FROM activities WHERE attempt=? ORDER BY rowid",
                    (current["id"],),
                )
            }
        )
        # A consumer's build phase establishes availability of its required
        # input outputs, without claiming local builds or successful checks.
        for r in db.execute(
            """SELECT e.outputs AS required,d.outputs FROM edges e
          JOIN activities a ON a.drv=e.parent JOIN derivations d ON d.drv=e.child
          WHERE a.attempt=? AND a.phase IS NOT NULL""",
            (current["id"],),
        ):
            required = json.loads(r["required"])
            if isinstance(required, dict):
                required = required.get("outputs", [])
            outputs = json.loads(r["outputs"])
            provided.update(outputs[k] for k in required if outputs.get(k))
        requested.update(json.loads(current["spec"]).get("derivations", []))
    # Restrict blockers and reverse edges to this campaign, including its native tools.
    scope = {r[0] for r in db.execute(SCOPE + "SELECT drv FROM scope", (campaign,))}
    bad = {
        r[0]
        for r in db.execute("""WITH RECURSIVE bad(drv) AS (
      SELECT drv FROM derivations WHERE failure IS NOT NULL OR exclusion IS NOT NULL
      UNION SELECT parent FROM edges JOIN bad ON child=bad.drv) SELECT drv FROM bad""")
    }
    cache = {}
    labels_by_drv = {}
    for r in db.execute(
        "SELECT id,label,drv FROM candidates WHERE campaign=? AND drv IS NOT NULL ORDER BY label",
        (campaign,),
    ):
        labels_by_drv.setdefault(r["drv"], []).append(
            dict(id=r["id"], label=r["label"])
        )

    def node(drv, required=None):
        if drv not in cache:
            row = db.execute("SELECT * FROM derivations WHERE drv=?", (drv,)).fetchone()
            cache[drv] = (
                dict(row)
                if row
                else dict(
                    drv=drv,
                    name=drv.rsplit("/", 1)[-1][33:-4],
                    outputs="{}",
                    available=0,
                    failure=None,
                    origin="unknown",
                )
            )
        record = cache[drv]
        outputs = json.loads(record["outputs"])
        wanted = required if required is not None else list(outputs)
        paths = [outputs.get(k) for k in wanted]
        present = bool(paths) and None not in paths and set(paths) <= existing
        consumed = (
            bool(paths) and None not in paths and set(paths) <= existing | provided
        )
        activity = activities.get(drv) if in_flight else None
        if activity and not activity["stopped"]:
            status = "building"
        elif record.get("exclusion"):
            status = "excluded"
        elif consumed:
            status = "available"
        elif activity:
            status = "settling"
        elif record["available"] or present:
            status = "available"
        elif record["failure"]:
            status = "failed"
        elif drv in bad:
            status = "blocked"
        elif drv in requested:
            status = "waiting"
        else:
            status = "unknown"
        labels = labels_by_drv.get(drv, [])[:8]
        return dict(
            drv=drv,
            name=record["name"],
            state=status,
            phase=activity["phase"] if activity else None,
            failure=record.get("exclusion") or record["failure"],
            origin="pre-existing" if present else record["origin"],
            availability_evidence="consumer-phase"
            if consumed and not present and not record["available"]
            else "store-observation"
            if status == "available"
            else None,
            labels=labels,
            required_outputs=wanted,
            attempt=activity["attempt"] if activity else record.get("evidence_attempt"),
        )

    roots = [node(d) for d in targets]
    building = [
        node(d)
        for d, a in activities.items()
        if in_flight and not a["stopped"] and d in scope
    ]
    if focus is None:
        focus = next((n["drv"] for n in building), None)
        focus = focus or next(
            (n["drv"] for n in roots if n["state"] in ("failed", "blocked")), None
        )
        focus = focus or next(
            (n["drv"] for n in roots if n["state"] != "available"), None
        )
        focus = focus or (targets[0] if targets else next(iter(sorted(scope)), None))
    elif focus not in scope:
        raise ValueError("derivation is outside this campaign's evaluated graph")

    work = dict(
        kind=active["kind"] if active else "idle",
        attempt=active["id"] if active else None,
        started=active["created"] if active else None,
    )
    planning = None
    if planner:
        total = len(json.loads(planner["targets"]))
        completed = 0
        try:
            with (directory(state, planner["id"]) / "plan.jsonl").open("rb") as f:
                completed = f.read(8 * 1024**2).count(b"\n")
        except OSError:
            pass
        planning = dict(
            attempt=planner["id"], completed=min(completed, total), total=total
        )
        if active["kind"] == "plan":
            work.update(completed=planning["completed"], total=total)
    if in_flight:
        work["stage"] = (
            "building" if building else "preflight" if before is None else "resolving"
        )
    result = dict(
        campaign=campaign,
        mode=c["mode"],
        heartbeat=c["heartbeat"],
        now=stamp(),
        work=work,
        planning=planning,
        batch=dict(id=batch["id"], state=batch["state"], created=batch["created"])
        if batch
        else None,
        batches=[
            dict(id=r["id"], state=r["state"], created=r["created"]) for r in batches
        ],
        roots=roots,
        building=building[:30],
        focus=None,
        inputs=[],
        consumers=[],
        totals={},
        selected_dependents=0,
        affected_roots=[],
        edges=[],
        page=page,
    )
    if focus is None:
        return result
    result["focus"] = node(focus)
    # Hubs can have thousands of consumers. Fetch their records in one indexed
    # query rather than issuing two SQL queries for every displayed/hidden node.
    for r in db.execute(
        """SELECT * FROM derivations WHERE drv IN (
      SELECT child FROM edges WHERE parent=? UNION SELECT parent FROM edges WHERE child=?)""",
        (focus, focus),
    ):
        cache[r["drv"]] = dict(r)
    input_roles = {}
    for r in db.execute(
        "SELECT child,role FROM roles WHERE campaign=? AND parent=? ORDER BY role",
        (campaign, focus),
    ):
        input_roles.setdefault(r["child"], []).append(r["role"])
    inputs, consumers = [], []
    for e in db.execute(
        "SELECT child,outputs FROM edges WHERE parent=? ORDER BY child", (focus,)
    ):
        required = json.loads(e["outputs"])
        if isinstance(required, dict):
            required = required.get("outputs", [])
        n = node(e["child"], required)
        n["roles"] = input_roles.get(e["child"], [])
        inputs.append(n)
    for e in db.execute(
        "SELECT parent FROM edges WHERE child=? ORDER BY parent", (focus,)
    ):
        if e[0] in scope:
            consumers.append(node(e[0]))
    priority = {
        "building": 0,
        "excluded": 1,
        "failed": 1,
        "blocked": 2,
        "settling": 3,
        "waiting": 4,
        "unknown": 5,
        "available": 6,
    }

    def order(n):
        return (priority[n["state"]], n["name"], n["drv"])

    inputs.sort(key=order)
    consumers.sort(key=order)
    hidden = 0 if show_available else sum(n["state"] == "available" for n in inputs)
    visible = (
        inputs if show_available else [n for n in inputs if n["state"] != "available"]
    )
    page = max(0, min(page, max(0, (max(len(visible), len(consumers)) - 1) // 6)))
    result["page"] = page
    result["inputs"] = visible[page * 6 : page * 6 + 6]
    result["consumers"] = consumers[page * 6 : page * 6 + 6]
    result["totals"] = dict(
        inputs=len(inputs),
        consumers=len(consumers),
        hidden_available=hidden,
        input_states=dict(Counter(n["state"] for n in inputs)),
        more=(page + 1) * 6 < max(len(visible), len(consumers)),
    )
    result["edges"] = [
        dict(source=n["drv"], target=focus) for n in result["inputs"]
    ] + [dict(source=focus, target=n["drv"]) for n in result["consumers"]]
    up = {
        r[0]
        for r in db.execute(
            """WITH RECURSIVE up(drv) AS (
      VALUES(?) UNION SELECT parent FROM edges JOIN up ON child=up.drv) SELECT drv FROM up""",
            (focus,),
        )
    }
    result["affected_roots"] = [
        n for n in roots if n["drv"] in up and n["drv"] != focus
    ]
    result["selected_dependents"] = sum(
        1
        for r in db.execute(
            "SELECT drv FROM candidates WHERE campaign=? AND drv IS NOT NULL",
            (campaign,),
        )
        if r[0] in up and r[0] != focus
    )
    return result
